### ISSUE TITLE
Potential Deadlock on production server

### ISSUE THREAD

**Issue by JamesPHoughton** [empiricaly/empirica#379]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Deadlock issue running server in production mode. I'm a few versions behind current on the server at the moment, so I'm going to upgrade and see if I still have the problem. Thought i'd put this here to document the issue.

### Empirica Version

```text
Version: v1.3.2
SHA:     a1c7ead
Branch:  main
Time:    2023-05-14T11:24:08Z

Client:  1.3.2
Server:  1.3.2
```

### What OS are you seeing the problem on?

Linux

### What browser are you seeing the problem on?

Does not apply

### Relevant log output

```shell
2023-07-31T06:45:59.946-04:00	POTENTIAL DEADLOCK:

2023-07-31T06:45:59.947-04:00	Previous place where the lock was grabbed

2023-07-31T06:45:59.947-04:00	goroutine 356922 lock 0xc00007d240

2023-07-31T06:45:59.947-04:00	go/pkg/mod/github.com/empiricaly/tajriba@v1.1.1/internal/runtime/change.go:400 runtime.(*Runtime).SubChanges.func1 ??? <<<<<

2023-07-31T06:45:59.947-04:00	go/pkg/mod/github.com/sasha-s/go-deadlock@v0.3.1/deadlock.go:116 go-deadlock.(*RWMutex).Lock ???

2023-07-31T06:45:59.947-04:00	Have been trying to lock it again for more than 30s

2023-07-31T06:45:59.947-04:00	goroutine 360389 lock 0xc00007d240

2023-07-31T06:45:59.947-04:00	go/pkg/mod/github.com/empiricaly/tajriba@v1.1.1/internal/runtime/hooks.go:172 runtime.(*Runtime).SubOnEvent.func1 ??? <<<<<

2023-07-31T06:45:59.947-04:00	go/pkg/mod/github.com/sasha-s/go-deadlock@v0.3.1/deadlock.go:116 go-deadlock.(*RWMutex).Lock ???

2023-07-31T06:45:59.955-04:00	Here is what goroutine 356922 doing now

2023-07-31T06:45:59.955-04:00	goroutine 356922 [chan send, 1 minutes]:

2023-07-31T06:45:59.955-04:00	github.com/empiricaly/tajriba/internal/runtime.(*Runtime).propagateHook(0xc00007d200, {0xc00047bda0?, 0xc0000d0300?}, {0xcc770d, 0x16}, {0xc0000d0300, 0x1a}, {0xde4820, 0xc00026dab0})

2023-07-31T06:45:59.955-04:00	/go/pkg/mod/github.com/empiricaly/tajriba@v1.1.1/internal/runtime/hooks.go:43 +0x407

2023-07-31T06:45:59.955-04:00	github.com/empiricaly/tajriba/internal/runtime.(*Runtime).SubChanges.func1()

2023-07-31T06:45:59.955-04:00	/go/pkg/mod/github.com/empiricaly/tajriba@v1.1.1/internal/runtime/change.go:417 +0x5de

2023-07-31T06:45:59.955-04:00	created by github.com/empiricaly/tajriba/internal/runtime.(*Runtime).SubChanges

2023-07-31T06:45:59.955-04:00	/go/pkg/mod/github.com/empiricaly/tajriba@v1.1.1/internal/runtime/change.go:378 +0x1ec

2023-07-31T06:45:59.955-04:00	Other goroutines holding locks:

2023-07-31T06:45:59.955-04:00	goroutine 356922 lock 0xc000d83c74

2023-07-31T06:45:59.955-04:00	go/pkg/mod/github.com/empiricaly/tajriba@v1.1.1/internal/runtime/hooks.go:41 runtime.(*Runtime).propagateHook ??? <<<<<

2023-07-31T06:45:59.955-04:00	go/pkg/mod/github.com/empiricaly/tajriba@v1.1.1/internal/runtime/hooks.go:40 runtime.(*Runtime).propagateHook ???

2023-07-31T06:45:59.955-04:00	go/pkg/mod/github.com/empiricaly/tajriba@v1.1.1/internal/runtime/change.go:418 runtime.(*Runtime).SubChanges.func1 ???

2023-07-31T06:45:59.957-04:00	/scripts/entrypoint.sh: line 18: wait: `': not a pid or valid job spec
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by npaton** [comment]
I think this was fixed around v1.3.4. If you can upgrade.

**Comment by JamesPHoughton** [comment]
Had the bug again during a run today 😩
```


2023-08-01T14:54:10.816-04:00 | POTENTIAL DEADLOCK:
-- | --
  | 2023-08-01T14:54:10.816-04:00 | Previous place where the lock was grabbed
  | 2023-08-01T14:54:10.816-04:00 | goroutine 96209 lock 0xc00007d0c0
  | 2023-08-01T14:54:10.816-04:00 | go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:380 runtime.(*Runtime).SubChanges.func1 ??? <<<<<
  | 2023-08-01T14:54:10.816-04:00 | go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:378 runtime.(*Runtime).SubChanges.func1 ???
  | 2023-08-01T14:54:10.816-04:00 | Have been trying to lock it again for more than 30s
  | 2023-08-01T14:54:10.816-04:00 | goroutine 96227 lock 0xc00007d0c0
  | 2023-08-01T14:54:10.816-04:00 | go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:380 runtime.(*Runtime).SubChanges.func1 ??? <<<<<
  | 2023-08-01T14:54:10.816-04:00 | go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:378 runtime.(*Runtime).SubChanges.func1 ???
  | 2023-08-01T14:54:10.945-04:00 | Here is what goroutine 96209 doing now
  | 2023-08-01T14:54:10.945-04:00 | goroutine 96209 [select]:
  | 2023-08-01T14:54:10.945-04:00 | github.com/empiricaly/tajriba/internal/runtime.(*changesSub).Send(0xc002932d40, {0xc00333e000, 0x8e, 0x1a?})
  | 2023-08-01T14:54:10.945-04:00 | /go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:55 +0x105
  | 2023-08-01T14:54:10.945-04:00 | github.com/empiricaly/tajriba/internal/runtime.(*changesSub).publish(0xc002932d40, {0xc00017a420?, 0xc001a2ae60?}, {0xc0023e1000, 0x8e, 0x100})
  | 2023-08-01T14:54:10.945-04:00 | /go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:357 +0x5bb
  | 2023-08-01T14:54:10.945-04:00 | github.com/empiricaly/tajriba/internal/runtime.(*Runtime).pushLinks(0xc00007d080, {0xdf2140, 0xc00534db60}, {0xc002932dc0, 0x6, 0xc00007d018?}, 0xc002932d40)
  | 2023-08-01T14:54:10.945-04:00 | /go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:285 +0x949
  | 2023-08-01T14:54:10.945-04:00 | github.com/empiricaly/tajriba/internal/runtime.(*Runtime).SubChanges.func1()
  | 2023-08-01T14:54:10.945-04:00 | /go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:390 +0x250
  | 2023-08-01T14:54:10.945-04:00 | created by github.com/empiricaly/tajriba/internal/runtime.(*Runtime).SubChanges
  | 2023-08-01T14:54:10.945-04:00 | /go/pkg/mod/github.com/empiricaly/tajriba@v1.2.0/internal/runtime/change.go:378 +0x1ec
  | 2023-08-01T14:54:10.954-04:00 | /scripts/entrypoint.sh: line 18: wait: `': not a pid or valid job spec

```

I'm running:

> version: v1.4.6
> sha: 877d1e0
> tag: v1.4.6
> branch: main
> time: "2023-07-16T15:01:49Z"
> 
>     "@empirica/core": "^1.5.0-next.3",

**Comment by npaton** [comment]
I released 1.5.0 that should help with this issue. But it is very difficult to validate the solution.

The problem seems to be about slow/zombie player connections. Updates (attribute changes, scope subs, group changes, etc.) are sent to players over websocket synchronously as they are made in the backend. The problem is if one connection is lagging, it can block others. Normally, this is not a problem since the connection has pings and should timeout relatively quickly. But with a flaky connection that is just surviving, intermittently slow, it can create a lock up. At least, that's my postulate.

The ultimate solution here would be make the player dispatch asynchronous. But that's a very large change. So right now, I've added a short timeout on the dispatch for each connection, then each connection fallbacks to a temporary buffer so it can pick back up if it was just a fluke, and finally we will abandon the connection and try to kill it if it keeps failing (though the killing of the connection is harder than you'd think and it's untested).

In any case, the deadlock in the backend should be removed. We have a short timeout for each connection to stay in the sync mode and if that fails, it's removed from the pool of sync connections. I am concerned how bad connections will behave, though all data is ordered, so there shouldn't be any invalid data pushed, and usually, at that point, the client is probably loosing connectivity and a restart of the connection should happen on its own.

I've tested this with a some automated integration tests. I was able to reproduce the problem and see it fixed in the tests, but it's complicated, and there's only so much I can do, so I have to release it out to see how it goes. Please let me know.

**Comment by JamesPHoughton** [comment]
Thanks @npaton, incorporated into our code and passing all our tests. We'll get it in front of an audience this week, and I'll let you know how it goes. =)
